src/main/java/hudson/plugins/promoted_builds/integrations/jobdsl/PromotionContext.java

```diff
@@ -1,108 +1,145 @@
 package hudson.plugins.promoted_builds.integrations.jobdsl;
 
+import edu.umd.cs.findbugs.annotations.SuppressFBWarnings;
 import groovy.lang.Closure;
+import groovy.lang.MetaClass;
 import groovy.util.Node;
-import hudson.plugins.promoted_builds.PromotionCondition;
-import hudson.plugins.promoted_builds.conditions.DownstreamPassCondition;
-import hudson.plugins.promoted_builds.conditions.ParameterizedSelfPromotionCondition;
-import hudson.plugins.promoted_builds.conditions.SelfPromotionCondition;
-import hudson.plugins.promoted_builds.conditions.UpstreamPromotionCondition;
+import groovy.util.NodeBuilder;
+import groovy.util.NodeList;
+import javaposse.jobdsl.dsl.DslContext;
+import javaposse.jobdsl.dsl.Item;
+import javaposse.jobdsl.dsl.JobManagement;
+import javaposse.jobdsl.dsl.helpers.step.StepContext;
+import javaposse.jobdsl.plugin.DslEnvironment;
+import org.codehaus.groovy.runtime.InvokerHelper;
+import org.codehaus.groovy.runtime.MethodClosure;
 
 import java.util.ArrayList;
 import java.util.List;
 
-import javaposse.jobdsl.dsl.Context;
-import javaposse.jobdsl.dsl.helpers.step.StepContext;
-
-import javaposse.jobdsl.plugin.DslEnvironment;
-
 import static javaposse.jobdsl.plugin.ContextExtensionPoint.executeInContext;
 
-class PromotionContext implements Context {
-    private final DslEnvironment dslEnvironment;
+/**
+ * Implementation notes: in order to get advanced integration with Job DSL which is Groovy-powered, we make use of
+ * some helper classes such as the {@link MethodClosure} adapter used for methods accepting Closure arguments. This may
+ * result in some unusual OOP constructs.
+ */
+class PromotionContext extends Item {
```

Comment:
You can use `@ContextType` and `AbstractExtensibleContext`. I`ll update the docs.

The "XML generation engine" is very lightweight. You do not need to extend `Item` for that, just add your own XML processing code. This is only relevant for the Promoted Builds plugin, so there will be no Job DSL API for this.

And extensions should not need to have their own configure blocks. An extension should provide DSL syntax for all necessary methods. The plugin's maintainer must be aware of the Job DSL extension and should require contributions to add Job DSL extension code. Your approach to use the Automatically Generated DSL will mitigate this problem.